nearcache_map.go

```diff
@@ -0,0 +1,265 @@
+	fmt.Println("IMPLEMENT ME: registerInvalidationListener")
+}
+
+func (ncm *nearCacheMap) preload() error {
+	panic("implement me!")
+}
+
+func (ncm *nearCacheMap) ContainsKey(ctx context.Context, key interface{}, m *Map) (found bool, err error) {
+	key, err = ncm.toNearCacheKey(key)
+	if err != nil {
+		return false, err
+	}
+	cached, ok, err := ncm.getCachedValue(key, false)
+	if err != nil {
+		return false, err
+	}
+	if ok {
+		return cached != nil, nil
+	}
+	keyData, err := m.validateAndSerialize(key)
+	if err != nil {
+		return false, err
+	}
+	return m.containsKeyFromRemote(ctx, keyData)
+}
+
+func (ncm *nearCacheMap) Delete(ctx context.Context, m *Map, key interface{}) error {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return err
+	}
+	defer ncm.nc.Invalidate(key)
+	keyData, err := m.validateAndSerialize(key)
+	if err != nil {
+		return err
+	}
+	return m.deleteFromRemote(ctx, keyData)
+}
+
+func (ncm *nearCacheMap) Get(ctx context.Context, m *Map, key interface{}) (interface{}, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return nil, err
+	}
+	cached, found, err := ncm.getCachedValue(key, false)
+	if err != nil {
+		return nil, err
+	}
+	if found {
+		return cached, nil
+	}
+	// value not found in local cache.
+	// get it from remote.
+	value, err := ncm.getFromRemote(ctx, m, key)
+	if err != nil {
+		ncm.nc.Invalidate(key)
+		return nil, err
+	}
+	return value, nil
+}
+
+func (ncm *nearCacheMap) Put(ctx context.Context, m *Map, key, value interface{}, ttl int64) (interface{}, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return nil, err
+	}
+	defer ncm.nc.Invalidate(key)
+	keyData, valueData, err := m.validateAndSerialize2(key, value)
+	if err != nil {
+		return false, err
+	}
+	return m.putWithTTLFromRemote(ctx, keyData, valueData, ttl)
+}
+
+func (ncm *nearCacheMap) PutWithMaxIdle(ctx context.Context, m *Map, key, value interface{}, ttl int64, maxIdle int64) (interface{}, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return nil, err
+	}
+	defer ncm.nc.Invalidate(key)
+	keyData, valueData, err := m.validateAndSerialize2(key, value)
+	if err != nil {
+		return false, err
+	}
+	return m.putWithMaxIdleFromRemote(ctx, keyData, valueData, ttl, maxIdle)
+}
+
+func (ncm *nearCacheMap) Remove(ctx context.Context, m *Map, key interface{}) (interface{}, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return false, err
+	}
+	defer ncm.nc.Invalidate(key)
+	keyData, err := m.validateAndSerialize(key)
+	if err != nil {
+		return nil, err
+	}
+	return m.removeFromRemote(ctx, keyData)
+}
+
+func (ncm *nearCacheMap) RemoveIfSame(ctx context.Context, m *Map, key interface{}, value interface{}) (bool, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return false, err
+	}
+	defer ncm.nc.Invalidate(key)
+	keyData, valueData, err := m.validateAndSerialize2(key, value)
+	if err != nil {
+		return false, err
+	}
```

Comment:
Similarly, you can move serialization logic to removeIfSameFromRemote

Reply:
Refactored at: https://github.com/hazelcast/hazelcast-go-client/pull/806/commits/a5b6adc0848f374f5049cf9a75bcb2aacccf9699